pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1542 @@
+// is observed.
+func (proxier *Proxier) OnNodeAdd(node *v1.Node) {
+	if node.Name != proxier.hostname {
+		klog.ErrorS(nil, "Received a watch event for a node that doesn't match the current node",
+			"eventNode", node.Name, "currentNode", proxier.hostname)
+		return
+	}
+
+	if reflect.DeepEqual(proxier.nodeLabels, node.Labels) {
+		return
+	}
+
+	proxier.mu.Lock()
+	proxier.nodeLabels = map[string]string{}
+	for k, v := range node.Labels {
+		proxier.nodeLabels[k] = v
+	}
+	proxier.mu.Unlock()
+	klog.V(4).InfoS("Updated proxier node labels", "labels", node.Labels)
+
+	proxier.Sync()
+}
+
+// OnNodeUpdate is called whenever modification of an existing
+// node object is observed.
+func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
+	if node.Name != proxier.hostname {
+		klog.ErrorS(nil, "Received a watch event for a node that doesn't match the current node",
+			"eventNode", node.Name, "currentNode", proxier.hostname)
+		return
+	}
+
+	if reflect.DeepEqual(proxier.nodeLabels, node.Labels) {
+		return
+	}
+
+	proxier.mu.Lock()
+	proxier.nodeLabels = map[string]string{}
+	for k, v := range node.Labels {
+		proxier.nodeLabels[k] = v
+	}
+	proxier.mu.Unlock()
+	klog.V(4).InfoS("Updated proxier node labels", "labels", node.Labels)
+
+	proxier.Sync()
+}
+
+// OnNodeDelete is called whenever deletion of an existing node
+// object is observed.
+func (proxier *Proxier) OnNodeDelete(node *v1.Node) {
+	if node.Name != proxier.hostname {
+		klog.ErrorS(nil, "Received a watch event for a node that doesn't match the current node",
+			"eventNode", node.Name, "currentNode", proxier.hostname)
+		return
+	}
+
+	proxier.mu.Lock()
+	proxier.nodeLabels = nil
+	proxier.mu.Unlock()
+
+	proxier.Sync()
+}
+
+// OnNodeSynced is called once all the initial event handlers were
+// called and the state is fully propagated to local cache.
+func (proxier *Proxier) OnNodeSynced() {
+}
+
+const (
+	// Maximum length of an nftables chain name. (FIXME: put this in danwinship/nftables)
+	nftablesChainNameLengthMax = 128
+
+	// Maximum length for one of our chain name prefixes, including the trailing
+	// hyphen.
+	chainNamePrefixLengthMax = 16
+
+	// Maximum length of the string returned from servicePortChainNameBase or
+	// servicePortEndpointChainNameBase.
+	chainNameBaseLengthMax = nftablesChainNameLengthMax - chainNamePrefixLengthMax
+)
+
+const (
+	servicePortPolicyClusterChainNamePrefix = "service-"
+	servicePortPolicyLocalChainNamePrefix   = "local-"
+	serviceExternalChainNamePrefix          = "external-"
+	servicePortEndpointChainNamePrefix      = "endpoint-"
+	servicePortEndpointAffinityNamePrefix   = "affinity-"
+)
+
+// servicePortChainNameBase returns the base name for a chain for the given ServicePort.
+func servicePortChainNameBase(servicePortName *proxy.ServicePortName, protocol string) string {
+	// nftables chains can contain the characters [A-Za-z0-9_./-]
+
+	// Namespace/Service/Port names are all, roughly /[a-z][a-z-]{0,62}/
```

Comment:
https://github.com/kubernetes/kubernetes/blob/86a0b1c57a1028729deb82d833007b20955a6cb3/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L60-L63

and 

https://github.com/kubernetes/kubernetes/blob/86a0b1c57a1028729deb82d833007b20955a6cb3/pkg/apis/core/validation/validation.go#L255

This is something we need to be completely sure or this can lead to subtle bugs

Reply:
Oops, I _did_ dig into the validation functions, but then I rewrote the comments a bunch of times (trying to decide how much detail to go into) and ended up with the wrong regexps. It should have said

`// Namespace/Service/Port names are all, roughly /[a-z0-9][a-z0-9-]{0,62}/`

(The difference between `NameIsDNSLabel` and `NameIsDNS1035Label` btw is that the latter requires an initial letter rather than an initial letter-or-number. And neither allows a trailing hyphen, but that doesn't matter for our purposes here.)

I should probably go into more detail in the comments to make it clearer to what extent you could tweak things in the future without breaking things...